src/platform_impl/windows/event_loop.rs

```diff
@@ -0,0 +1,1603 @@
+    pub mouse_buttons_down: u32,
+}
+
+impl WindowState {
+    pub fn update_min_max(&mut self, old_dpi_factor: f64, new_dpi_factor: f64) {
+        let scale_factor = new_dpi_factor / old_dpi_factor;
+        let dpi_adjuster = |mut physical_size: PhysicalSize| -> PhysicalSize {
+            physical_size.width *= scale_factor;
+            physical_size.height *= scale_factor;
+            physical_size
+        };
+        self.max_size = self.max_size.map(&dpi_adjuster);
+        self.min_size = self.min_size.map(&dpi_adjuster);
+    }
+}
+
+pub(crate) struct SubclassInput<T> {
+    pub window_state: Arc<Mutex<WindowState>>,
+    pub event_loop_runner: EventLoopRunnerShared<T>,
+    pub file_drop_handler: FileDropHandler,
+}
+
+impl<T> SubclassInput<T> {
+    unsafe fn send_event(&self, event: Event<T>) {
+        self.event_loop_runner.send_event(event);
+    }
+}
+
+struct ThreadMsgTargetSubclassInput<T> {
+    event_loop_runner: EventLoopRunnerShared<T>,
+    user_event_receiver: Receiver<T>
+}
+
+impl<T> ThreadMsgTargetSubclassInput<T> {
+    unsafe fn send_event(&self, event: Event<T>) {
+        self.event_loop_runner.send_event(event);
+    }
+}
+
+pub struct EventLoop<T> {
+    // Id of the background thread from the Win32 API.
+    thread_id: DWORD,
+    thread_msg_target: HWND,
+    thread_msg_sender: Sender<T>,
+    trigger_newevents_on_redraw: Arc<AtomicBool>,
+    pub(crate) runner_shared: EventLoopRunnerShared<T>,
+}
+
+impl<T: 'static> EventLoop<T> {
+    pub fn new() -> EventLoop<T> {
+        Self::with_dpi_awareness(true)
+    }
+
+    pub fn with_dpi_awareness(dpi_aware: bool) -> EventLoop<T> {
+        become_dpi_aware(dpi_aware);
+
+        let thread_id = unsafe { processthreadsapi::GetCurrentThreadId() };
+        let runner_shared = Rc::new(ELRShared {
+            runner: RefCell::new(None),
+            buffer: RefCell::new(VecDeque::new())
+        });
+        let (thread_msg_target, thread_msg_sender) = thread_event_target_window(runner_shared.clone());
+
+        EventLoop {
+            thread_id,
+            thread_msg_target, thread_msg_sender,
+            trigger_newevents_on_redraw: Arc::new(AtomicBool::new(true)),
+            runner_shared
+        }
+    }
+
+    pub fn run<F>(mut self, event_handler: F) -> !
+        where F: 'static + FnMut(Event<T>, &RootEventLoop<T>, &mut ControlFlow)
+    {
+        self.run_return(event_handler);
+        ::std::process::exit(0);
+    }
+
+    pub fn run_return<F>(&mut self, mut event_handler: F)
+        where F: FnMut(Event<T>, &RootEventLoop<T>, &mut ControlFlow)
+    {
+        unsafe{ winuser::IsGUIThread(1); }
+
+        assert_eq!(mem::size_of::<RootEventLoop<T>>(), mem::size_of::<EventLoop<T>>());
+        let self_ptr = self as *const EventLoop<T>;
+
+        let mut runner = unsafe{ EventLoopRunner::new(
+            self,
+            move |event, control_flow| {
+                let event_loop_ref = &*(self_ptr as *const RootEventLoop<T>);
```

Comment:
This kind of trick cannot be done for x11/wayland (as there is an internal layer of `EventLoop` type that is an enum and does the dispatch between x11 and wayland). As such, I'm not sure how I can implement `run_return` for wayland without some seriously UB black magic.

Reply:
The original design added a new `EventLoopWindows` type used for window creation that got passed to the callback where `EventLoop` currently gets passed. I removed it because I realized you could pass a immutable reference to `EventLoop` to achieve the same effect, but it may make sense to add that type back in to reduce the amount of unsafe shenanigans that have to take place.

Reply:
I think that'd be better, yes.

Reply:
I've re-added that type as `EventLoopWindowTarget`.

Reply:
Ah, nevermind, I missed some subtlety, I can work. Won't be extremely pretty, but that's the fault of the 2-in-1 x11/Wayland backend